Secrets 🔐 PR 6 - Update Oracle properties

jeff303 opened the pull request:
Adding keystore and truststore secret properties to Oracle YAML configuration

For more context see epic: #18036

jeff303 committed:
Add secret model

Add Liquibase migration for secret table

Implement bare bones model namespace for secret

Add simple test to ensure secret values are stored and retrieved under an encryption key (or not)

jeff303 committed:
Support encryption of binary data

Update encryption code to be able to operate on byte arrays directly (without dealing in base64 String representation), delegating existing methods to point to these new ones

Adding test

jeff303 committed:
Integrate secret handling with db-details

Add helper function to database model namespace to "wwap out" secret values with inserted IDs

Calling this from the pre-update fn

Add pre-insert fn for database model to perform the same

Adding test to database model test ns

TODO: add test for simply not touching one (ensure the existing data remains)

jeff303 committed:
Merging master back in to pick up changes

Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Add backend support for keystore type secrets

Adding code to handle keystore type secrets

Adding test code to create and verify a keystore through the secret DB roundtrip

TODO: change mentions of "jks" to "keystore", add support for source=file, add helper fn for secret->temp-file

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
More refinement of keystore type secrets

Removing mention of JKS and keeping "keystore" terminology

Fixing construction of KeyStore under JDK8

Adding util fn to secret namespace for converting the value to file, and adding test for that

jeff303 committed:
Fix support for source

jeff303 committed:
Add :creator_id to secret model to capture the user who created this secret instance/version

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Fix created_at and updated_at timestamp column types

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Add creator-id and created-at sub properties for each secret in db details

Update test to check for these

This should also be sufficient to break the connection details hash (PR #17307)

jeff303 committed:
Merge secrets-3 branch back into secrets-4 and fix conflicts

jeff303 committed:
Add missing remarks to make migrations linter happy

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Fix namespace decl for database_test.clj

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Fix failing tests

jeff303 committed:
PR feedback: move comment

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Support updating secret values as part of encryption key rotation

Add test assertions for secret values in the `rotate-encryption-key!-test`

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Switch to looking at a `-path` suffix for a secret property, to indicate that the secret should have :source :file-path

Update database model code and test accordingly

Fixing unrelated indentation issue in email_test.clj

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Fix clj-kondo error

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Translate secret connection-properties into existing types understood by frontend

For all :type :secret properties, before returning over the wire (i.e. from `available-drivers-info`), translate the connection-property into one or more new client side properties, using the existing types, to encapsulate the behavior of that individual secret "unit".  Depending on the :secret-kind and hosting environment, this could mean there are multiple additional properties injected (one for the file upload versus local path option field, one for the value itself on upload, and one for the file path)

Adding utility function to do the reverse (i.e. translating those client side properties back to the server secret property definition)

Adding tests for both (including updating secret-test-driver to cover all of these)

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Fix test failures

Add support for the treat-before-posting connection property config

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Add SSL keystore and truststore secret properties to Oracle driver YAML files

jeff303 committed:
Copy :visible-if from the secret prop to the emitted options prop

jeff303 committed:
Adding test for Oracle connection properties

Adding new `select-keys-sequentially` helper function for dealing with assertions on transformed conn props

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Merge branch 'secrets-4-backend-support-jks-file' into secrets-5-transform-connection-properties

jeff303 committed:
Merge branch 'secrets-5-transform-connection-properties' into secrets-6-update-oracle-properties

gusaiani committed:
Read as data

jeff303 committed:
Merge branch 'master' into secrets-1-backend-support-binary-encryption

jeff303 committed:
Merge branch 'secrets-1-backend-support-binary-encryption' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Merge branch 'secrets-4-backend-support-jks-file' into secrets-5-transform-connection-properties

jeff303 committed:
Merge branch 'secrets-5-transform-connection-properties' into secrets-6-update-oracle-properties

jeff303 committed:
Update Oracle driver to set keystore and truststore connection options from secret values

Add new function to the secret namespace to return a lighter weight "secret map", and remove that functionality from the `handle-db-details-secret-prop!`, so that it can be reused from connection testing

Modifying CircleCI to set the corresponding truststore secret conn prop vars instead of JAVA_OPTS level truststore settings

jeff303 committed:
PR feedback

jeff303 committed:
Merge branch 'master' into secrets-2-backend-add-secret-model

jeff303 committed:
Merge branch 'secrets-2-backend-add-secret-model' into secrets-3-integrate-db-details

jeff303 committed:
Merge branch 'secrets-3-integrate-db-details' into secrets-4-backend-support-jks-file

jeff303 committed:
Merge branch 'secrets-4-backend-support-jks-file' into secrets-5-transform-connection-properties

jeff303 committed:
Merge branch 'secrets-5-transform-connection-properties' into secrets-6-update-oracle-properties

jeff303 committed:
Work more on SSL connectivity for Oracle with secrets

Expand Oracle test connection details to incorporate all SSL related properties, and also make it a function instead of delayed def (so changes can be picked up later)

Misc. fixes in Oracle driver YAML files

Update test to actually run properly, and then test connectivity and then sync/query in succession

TODO: fix the sync/query functionality (some weird state issue, where the second call to driver/describe-database returns no tables)

Update `exception-classes-not-to-retry` to include SSLHandshakeException

Fix misc issues in secret and database model code

Update CircleCI config to use correct secret based env var keys and values

jeff303 committed:
Add *database-name-override* dynamic var in `metabase.test.data.interface` to allow for DB name overriding

Fix up ssl connectivity test so all parts pass successfully

jeff303 committed:
Get rid of crazy with-redefs to swap in the existing test-data's schema, and instead, just dropping the DB if it was created under a different schema (similar to what H2 does)

Make linters happy

Add sanity check in CircleCI

jeff303 committed:
Fix issues with running SSL test

jeff303 committed:
Fix env var name in CircleCI

jeff303 committed:
Merge branch 'master' into secrets-4-backend-support-jks-file

jeff303 committed:
Update added version strings

jeff303 committed:
Merge branch 'secrets-4-backend-support-jks-file' into secrets-5-transform-connection-properties

jeff303 committed:
Merge branch 'secrets-5-transform-connection-properties' into secrets-6-update-oracle-properties

jeff303 committed:
Add more debug output for test failure

jeff303 committed:
Fix table count check more

jeff303 committed:
Merge branch 'master' into secrets-4-backend-support-jks-file

jeff303 committed:
Merge branch 'secrets-4-backend-support-jks-file' into secrets-5-transform-connection-properties

jeff303 committed:
Fix namespace decls

jeff303 committed:
Merge branch 'secrets-5-transform-connection-properties' into secrets-6-update-oracle-properties

jeff303 committed:
Merge branch 'master' into secrets-5-transform-connection-properties

jeff303 committed:
Fix bug in `delete-orphaned-secrets!`

jeff303 committed:
Merge branch 'secrets-5-transform-connection-properties' into secrets-6-update-oracle-properties

jeff303 committed:
Merge branch 'master' into secrets-6-update-oracle-properties

jeff303 committed:
Fix namespace decls

jeff303 committed:
Fix awful typo that broke everything

jeff303 committed:
Remove tru

jeff303 committed:
Merge branch 'master' into secrets-6-update-oracle-properties

dpsutton commented on src/metabase/models/secret.clj:
```diff
@@ -32,29 +33,38 @@
       (if (.exists existing-file)
         existing-file
-        (throw (ex-info (tru "Secret {0} points to non-existent file: {1}" id secret-val)
+        (throw (ex-info (tru "Secret {0} points to non-existent file: {1}" (if id id "") secret-val)
```
`(or id "")`

dpsutton commented on src/metabase/models/secret.clj:
```diff
@@ -32,29 +33,38 @@
                            (.setReadable true true)
                            (.deleteOnExit))]
-      (log/tracef "Creating temp file for secret %d value at %s" id (.getAbsolutePath tmp-file))
+      (log/tracef "Creating temp file for secret %s value at %s" (if id id "") (.getAbsolutePath tmp-file))
```
`(or id "")`

dpsutton commented on test/metabase/test/data/interface.clj:
```diff
@@ -190,7 +198,14 @@
+      (->>
+        (take-last 30)
+        (apply str)))
+  #_(apply str (take-last 30 (str/replace (str/lower-case (str (or *database-name-override* database-name) \_ table-name)) #"-" "_"))))
```
should we get rid of the commented out code?

jeff303 committed:
Merge branch 'master' into secrets-6-update-oracle-properties

jeff303 committed:
PR feedback

dpsutton commented on modules/drivers/oracle/test/metabase/test/data/oracle.clj:
```diff
@@ -77,6 +89,38 @@
+(defn- destroy-test-database-if-created-in-different-session
+  "For Oracle, we have a randomly selected `session-schema`, in order to allow different test runs to proceed
+  independently. If we happen to have a test-database already in our app DB, that was created under a different session,
+  then we should just destroy it, since we won't be able to resync it anyway."
```
is it not better to have the creator clean up after themselves? I'm not following 100% how this issue arises.

jeff303 replied:
This line was written quite a while ago, now, but I had a fairly long Slack discussion with @camsaul  about it.  I'll find the context there and update the docstring accordingly.

jeff303 committed:
PR feedback

* more detailed explanation of `destroy-test-database-if-created-in-different-session`
* simplifying ssl property iteration

jeff303 commented on modules/drivers/oracle/test/metabase/test/data/oracle.clj:
```diff
@@ -30,16 +34,24 @@
+                  :password (tx/db-test-env-var-or-throw :oracle :password)
+                  :sid      (tx/db-test-env-var-or-throw :oracle :sid)
+                  :ssl      (tx/db-test-env-var :oracle :ssl false)}]
+    (loop [details details*
```
Done

gusaiani committed:
Treat base64 form inputs

This PR tackles one of the "form widgets" - the one called "FormTextFileWidget".

It adds one form of treatment for its data, converting it to base64.
